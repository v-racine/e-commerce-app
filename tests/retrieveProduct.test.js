const { Config } = require('../src/config/config');

Config.Get(process.env);

const AppFactory = require('../src/app');
const request = require('supertest');
const { signupHelper } = require('./testUtils/signupHelper');

describe('retrieve a product to edit', () => {
  const mockProductsRepo = {};
  const mockUsersRepo = {};

  const app = AppFactory({
    productsRepo: mockProductsRepo,
    usersRepo: mockUsersRepo,
  });

  describe('when: the db does not contain the product requested', () => {
    let rsp;

    let id = '12345';

    beforeEach(async () => {
      mockProductsRepo.getOne = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(undefined);
        });
      });

      const cookie = await signupHelper(mockUsersRepo, app);

      rsp = await request(app).get(`/admin/products/${id}/edit`).set('Cookie', cookie).send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we respond with "Product not found"', async () => {
      const text = rsp.text;
      expect(text).toBe('Product not found');

      expect(mockProductsRepo.getOne).toHaveBeenCalledWith(id);

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: the user wants to retrieve a product to edit but is not signed in', () => {
    let rsp;

    beforeEach(async () => {
      mockProductsRepo.getOne = jest.fn();

      rsp = await request(app).get('/admin/products/1/edit').send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we redirect to sign-in page', async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      expect(mockProductsRepo.getOne).toHaveBeenCalledTimes(0);

      const status = rsp.status;
      expect(status).toBe(302);
    });
  });

  describe('when: the user wants to retrieve a product to edit', () => {
    let rsp;

    let id = '1234';

    let product = {
      id: id,
      title: 'product1',
      price: '12.00',
    };

    beforeEach(async () => {
      mockProductsRepo.getOne = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(product);
        });
      });

      const cookie = await signupHelper(mockUsersRepo, app);

      rsp = await request(app).get(`/admin/products/${id}/edit`).set('Cookie', cookie).send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return a pre-filled form of the product', async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      expect(mockProductsRepo.getOne).toHaveBeenCalledWith(id);

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });
});
