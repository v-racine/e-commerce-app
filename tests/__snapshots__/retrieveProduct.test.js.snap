// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retrieve a product to edit when: the user wants to retrieve a product to edit but is not signed in then: we redirect to sign-in page 1`] = `"Found. Redirecting to /signin"`;

exports[`retrieve a product to edit when: the user wants to retrieve a product to edit then: we return a pre-filled form of the product 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
    <form method="POST">
      <input name="title" value="product1" />
      
      <input name="price" value="12.00" />
      
      <input name="image" type="file" />
      <button>Submit</button>
    </form>
  
        </div>
      </body>
    </html>
  "
`;
